### PR TITLE
associated type Span, Tracer as short-hand, and *Protocol types

### PR DESCRIPTION
**Pull request by ktoso**
**Motivation:**

This is a revival of #84
where we try to KEEP compatibility with versions below 5.7 with a
compatibility "legacy" tracer type, but otherwise move towards requiring
5.7 for all the "nice" apis that use associated types and `any
  TracerProtocol` and friends

Replaces https://github.com/apple/swift-distributed-tracing/pull/92

**Modifications:**

- `Tracer` -> `TracerProtocol`
- `Tracer` is now a namespace in order to `Tracer.withSpan {}` easily
- `Span` -> `SpanProtocol`
- Introduce `LegacyTracerProtocol` which does not make use of associated
  type Span, and can be used in 5.6 libraries; they can deprecate and
  move away form it ASAP as they start requiring 5.7

**Result:**

Offer the APIs we want in 5.7 but remain compatible with 5.6 until we
drop it as soon as 5.9 is released as stable - this allows us to adopt
eagerly in libraries without having to wait for 5.9 to drop.

### CODE REVIEW

**Review by stevapple — APPROVED** [review]
Really clever solution!

File: README.md [hunk 1]
```diff
@@ -261,7 +261,7 @@ When instrumenting server applications there are typically three parties involve
 
 1. [Application developers](#application-developers-setting-up-instruments) creating server-side applications
 2. [Library/Framework developers](#libraryframework-developers-instrumenting-your-software) providing building blocks to create these applications
-3. [Instrument developers](#instrument-developers-creating-an-instrument) providing tools to collect distributed metadata about your application
+3. [InstrumentProtocol developers](#instrument-developers-creating-an-instrument) providing tools to collect distributed metadata about your application
```
[the remark this answers is not in the document]
Comment (ktoso) [reply README.md:264]: Thanks yeah will clean it up

**Review by FranzBusch — APPROVED** [review]
LGTM!

**Comment by ktoso** [comment]
Pinged some folks for more review -- if you spot anything off please comment and I'll address in follow ups! 👍

**Comment by ktoso** [comment]
Resolves https://github.com/apple/swift-distributed-tracing/issues/78

**Pull request merged**